### ISSUE TITLE
aria-required-parent false positive for native <option role="option"> inside <select>

### ISSUE THREAD

**Issue by jnurthen** [dequelabs/axe-core#5351]
### Product

axe-core

### Product Version

4.12.1

### Latest Version

- [x] I have tested the issue with the latest version of the product

## Expectation
Native `<option>` elements inside a native `<select>` should not fail aria-required-parent, even when they include an explicit `role="option"`.

Browsers already expose options through the native control’s accessibility tree (combobox with internal/slotted listbox ownership). The required parent context for option is provided by the UA, not by a DOM ancestor with `role="listbox"` or `role="group"`.

## Actual
axe reports aria-required-parent (critical) on each `<option role="option">` inside a single-select `<select>`.

Axe maps single-select `<select>` to implicit role combobox, then treats explicit `role="option"` as requiring a DOM parent of listbox or group. Walking the DOM parent chain finds combobox, so the rule fails.

Removing the redundant `role="option"` attributes makes the violation disappear, which shows the failure is driven by the explicit role + DOM walk, not by a real missing listbox relationship in the accessibility tree.

## How to Reproduce
Fails:

```
<!DOCTYPE html>
<html lang="en">
  <body>
    <select aria-label="Select a timeframe">
      <option value="current" role="option" selected>Current</option>
      <option value="previous" role="option">Previous</option>
    </select>
  </body>
</html>
```
Result with axe-core 4.12.1: 2 × aria-required-parent on the <option> nodes, required parents reported as group, listbox.

Passes (same markup without explicit roles):

```
<!DOCTYPE html>
<html lang="en">
  <body>
    <select aria-label="Select a timeframe">
      <option value="current" selected>Current</option>
      <option value="previous">Previous</option>
    </select>
  </body>
</html>
```
Result: no aria-required-parent violations.

## Additional context
* Explicit `role="option"` on native `<option>` is redundant / generally invalid authoring (allowedRoles: false), but classifying it as a missing required parent is still a false positive for native `<select>`.
* Suggested fix: skip aria-required-parent for `<option>` / `<option role="option">` when owned by a native `<select>`, or treat native `<select>` as providing the required listbox context for its options.
* The rule should continue to catch orphan custom `role="option"` widgets that are not inside a real listbox / group.

**Comment by straker** [comment]
Thanks for the issue. I see you already created https://github.com/w3c/aria/issues/2896 which is the issue causing the problem here as well. Thanks for filing that.

**Comment by modusensus** [comment]
Hi! I'd like to pick this up — I've been looking at the code and have a proposed approach.

Quick recap of the root cause I found in `lib/checks/aria/aria-required-parent-evaluate.js`: `ariaRequiredParentEvaluate` walks the DOM parent chain (via `getMissingContext`) looking for the required parent roles. A native `<select>` maps to `combobox`, and `combobox` isn't in `option`'s required context (`listbox`/`group`), so the walk always fails — even though the UA already provides the option/listbox relationship in the accessibility tree.

Proposed fix (scoped to avoid regressions):
- Only skip the check for native `<option>` elements whose owning element is a native `<select>` (including when they carry an explicit redundant `role="option"`).
- Custom `role="option"` widgets that aren't inside a real `listbox`/`group` should still be reported, so the check's coverage is unchanged.

Two questions before I open a PR:
1. Placement — would you prefer the exception in the check's evaluate function, or in the rule's matches function (`aria-required-parent-matches`)? I'd lean toward the evaluate function so the logic stays co-located with the required-context walk, but happy to match the project's conventions.
2. Since w3c/aria#2896 is still open — do you want to wait for the spec resolution, or is a pragmatic axe-side fix acceptable now?

Happy to open a PR either way.

**Comment by modusensus** [comment]
Hi! I'd like to pick this up — I've been looking at the code and have a proposed approach.

Quick recap of the root cause I found in `lib/checks/aria/aria-required-parent-evaluate.js`: `ariaRequiredParentEvaluate` walks the DOM parent chain (via `getMissingContext`) looking for the required parent roles. A native `<select>` maps to `combobox`, and `combobox` isn't in `option`'s required context (`listbox`/`group`), so the walk always fails — even though the UA already provides the option/listbox relationship in the accessibility tree.

Proposed fix (scoped to avoid regressions):
- Only skip the check for native `<option>` elements whose owning element is a native `<select>` (including when they carry an explicit redundant `role="option"`).
- Custom `role="option"` widgets that aren't inside a real `listbox`/`group` should still be reported, so the check's coverage is unchanged.

Two questions before I open a PR:
1. Placement — would you prefer the exception in the check's evaluate function, or in the rule's matches function (`aria-required-parent-matches`)? I'd lean toward the evaluate function so the logic stays co-located with the required-context walk, but happy to match the project's conventions.
2. Since w3c/aria#2896 is still open — do you want to wait for the spec resolution, or is a pragmatic axe-side fix acceptable now?

Happy to open a PR either way.

**Comment by modusensus** [comment]
Following up on my earlier comment with an answer to my own first question.

**Placement:** `lib/rules/aria-required-parent-matches.js` is a six-line eligibility filter — `return !!requiredContext(role)` — so the exception belongs in `aria-required-parent-evaluate.js`, next to the required-context walk, rather than in `matches`. Unless you'd prefer it in `matches`, I'll go that way.

**Scope,** so it stays predictable either way: only native `<option>` elements whose owning element is a native `<select>` are skipped, including when they carry a redundant explicit `role="option"`. Custom `role="option"` widgets that aren't inside a real `listbox`/`group` keep failing, so coverage for ARIA widgets doesn't change.

That leaves one question before I open a PR: `w3c/aria#2896` is still open — do you want to wait for the spec resolution, or is a pragmatic axe-side fix acceptable in the meantime?
